### ISSUE TITLE
Disabling Cache

### ISSUE THREAD

**Issue by YOzaz** [bradcornford/Setter#19]
Hi, I need a way to disable cache. Globally, or as a param in `get()` method. Explanation below.
I'm running a daemon worker, which may be idle for a while. At some point, database connection may drop, and daemon reconnects to it if required.
However, your package stores values to cache, and therefore initial `get()` may store empty value as connection is broken. Therefore even if I reconnect at some point and query key again, value doesn't refresh on `get()` method call.

**Comment by bradcornford** [comment]
Hi, yes this does make sense. I added a cache overall to reduce the cost of hitting a database constantly. However your use case does show that the ability to toggle the cache would be useful. I'll get this implemented.

**Comment by bradcornford** [comment]
Okay, i've implemented this functionality on the develop branch if you want to check it out and have a go. I'll merge this to master, and version it, once your happy with the functionality.

**Comment by YOzaz** [comment]
Looks great! I've added some comments on your commit.
Can you also add a way to disable cache just for one `get()` request? As in my case, I don't need a cache to be disabled globally in general; just in one specific constructor I need to ensure data is fresh.
It can be a param in function call, or `$options` array which will overwrite general options temporary.

**Comment by bradcornford** [comment]
I've updated develop again with your suggestions. I have also made the options enableCache and disableCache to return the current Setter instance. This way you can achieve what you need by doing:

``` php

Setting::disableCache()->get('test');
Setting::enableCache();

```

**Comment by YOzaz** [comment]
That's correct, however, consider this (following your example):
* Value is retrieved normally, cache is set. Let's say, value is errorneous.
* Cache is disabled with `disableCache()`. Value is retrieved as "fresh" - all fine.
* Cache is enabled. Trying to retrieve value - got errorneous because this doesn't set it back to cache: 
https://github.com/bradcornford/Setter/blob/8d8ea9caf52c133636e53579cd1e53e0b1a27824/src/Cornford/Setter/SettingBase.php#L287

In other words: disabling cache is not the same as forcing value to refresh.

**Comment by bradcornford** [comment]
Okay, so would you prefer a cacheGet($key, $default), cacheSet($key, $value), and a refreshCache($key) so you can manually refresh an item? Or an extra parameter to to the get($key, $value, $refreshCache) method to force cache refresh as a boolean?

**Comment by YOzaz** [comment]
* Settings::uncached()->get('key') - this would be the same mechanics as in Laravel soft deletes; it would set a temporary cache check skip (through protected flag e.g.) and would force value retrieval from DB, but would still set it to cache (if it's enabled globally).
* Or Settings::fresh('key') - again, it would skip cache check, but would set a value if caching is enabled.

First option looks nicer.

**Comment by bradcornford** [comment]
Okay, i've implemented your first suggestion. I've allowed it to only work for a single get request. It will also re-cache the item if caching is enabled.

**Comment by YOzaz** [comment]
Function `uncached` should `return $this;`.
Also, interface has `recache` contract... a mistake?
Other than that - looks great! Thanks!

**Comment by bradcornford** [comment]
Okay, i've fixed those issues, and amended the last commit.

**Comment by YOzaz** [comment]
@bradcornford by any chance, will you port this to 1.x version as well?
I'm using Laravel 4.2 at the moment, and can't upgrade easily...

**Comment by bradcornford** [comment]
@YOzaz I've now ported this over to version 1, under the version 1.6.0. I've not had time to fully test the items, however the PHPSpec tests seem to be passing, so let me know how you get on.

**Comment by YOzaz** [comment]
Thanks! There's a planned maintenance at the end of this week at my side, therefore package should be updated as well. I'll let you know how it goes.

**Comment by YOzaz** [comment]
OK, updated package in DEV server and got an exception:
```
Cornford \ Setter \ Exceptions \ SettingArgumentException
Cache is required in boolean format.
```
... which is fine, as I didn't publish a config. However, I think it should fall into default config values, no?
Also, it would be good to mention that config mush be published in docs as well:
```
php artisan config:publish cornford/setter
```

**Comment by YOzaz** [comment]
SettingBase constructor has an array at the end - which means, package can't be used as a dependency injection anymore :(

**Comment by bradcornford** [comment]
I may change the contructor to check isset and type check, rather than not isset.

How are you using the package?

**Comment by YOzaz** [comment]
I was injecting it into my service, but changed it now to app('setting').
This needs fixing: https://github.com/bradcornford/Setter/blob/v1.6.0/src/Cornford/Setter/SettingServiceProvider.php#L33 - returns empty array...

**Comment by bradcornford** [comment]
Yeah, i've pushed fixes for config, and service provider.

**Comment by YOzaz** [comment]
Alright, I think it works. Let's see how deamon workers will behave now in long term.
Meanwhile - a minor - link to L4 version in master docs should point to 1.6.0.
https://github.com/bradcornford/Setter/blob/master/README.md

**Comment by YOzaz** [comment]
Oh by the way, I didn't test these methods - just disabled whole cache in config for speed. Will play with methods later on, at the end of the month.

**Comment by bradcornford** [comment]
Okay, by then i should have given it a once over myself too.
